docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,1171 @@
+# Open Enclave Sealing API
+
+This document describes API functions provided by the Open Enclave SDK for
+sealing/unsealing data against an enclave's identity.
+
+## Motivation
+
+Sealing is an important capability of TEEs, which allows an enclave to encrypt
+and/or integrity-protect data at rest, using keys (aka., sealing keys) derived
+from the enclave's own identity. TEEs may have distinct formulas for key
+derivation, and may support different key lengths. And that leads to the desire
+for a TEE-agnostic sealing API.
+
+### Objectives
+
+The sealing API should be:
+
+- *TEE-agnostic* - The API should hide TEE specifics to allow reuse of source
+  code across TEEs.
+- *Accommodative* - Should the developers choose to, the API should allow
+  explicit uses of TEE specific features. For example, in the case of SGX, the
+  API should allow deriving keys of type `PROVISION_SEAL_KEY`, which may not
+  have an equivalent on other TEEs.
+- *Easy to use* - It's always a challenge for average developers to encrypt
+  data securely by using seal keys directly. The API should shield
+  cryptographic complexities from developers, by offering comprehensive
+  protection through an intuitive interface.
+- *Interoperable with existing SDKs* - The Intel SGX SDK also provides sealing
+  capabilities. It's desirable for enclaves, regardless of the SDKs they built
+  with, to be able to exchange sealed blobs, when allowed by the policy.
+
+### Non-Objectives
+
+- Cross-device sealing/unsealing is **not** supported - Sealed blobs created on a device must be unsealed on that same device.
+- Cross-TEE sealing/unsealing is **not** supported - Sealed blobs created by an
+  enclave in one TEE (e.g., SGX) cannot be unsealed by another enclave in a
+  different TEE (e.g., OP-TEE), even if they were both signed by the same
+  private key.
+
+## User Experience
+
+At the minimum, two API functions are necessary, namely `oe_seal()` and
+`oe_unseal()`. The former encrypts user data into a blob (whose format is
+implementation specific) while the latter decrypts/verifies the given blob and
+returns the data back to the user.
+
+`oe_seal()` requires a seal key to do its job. But how can a seal key be passed to
+`oe_seal()`? Here are 3 options:
+
+1. `oe_seal()` takes the seal key directly as a function argument; or
+2. `oe_seal()` takes a `key_info` as its argument for deriving the desired seal
+   key; or
+3. `oe_seal()` takes a `seal_policy` as its argument, and calls
+   `oe_get_seal_key_by_policy()` to obtain the seal key.
+
+Option 1 implies access to the seal key by software, which isn't always true
+(e.g., TPM never reveals seal keys to software). Option 3 prohibits developers
+from fine-tuning `key_info` (probably in a TEE-specific manner), which violates
+the *Accommodative* requirement. Option 2 is our choice.
```

Comment:
Which option covers the case where you pass in TEE-agnostic fields (like the seal policy), plus optionally a pointer to a TEE-specific struct?   That's the option that would be my choice.